### PR TITLE
Currency Service

### PR DESCRIPTION
**Pull request by brboi**

### CODE REVIEW

File: addons/wowl/static/src/utils/misc.ts [hunk 1]
```diff
@@ -53,7 +53,7 @@ export function json_node_to_xml(node: any, human_readable: any, indent: number)
     <any>!node.children instanceof Array ||
     <any>!node.attrs instanceof Object
   ) {
-    throw new Error(`"Node [%s] is not a JSONified XML node" ${JSON.stringify(node)}`);
+    throw new Error(`"Node [${JSON.stringify(node)}] is not a JSONified XML node"`);
```
Comment (brboi) [inline addons/wowl/static/src/utils/misc.ts]: remove "

File: addons/wowl/static/src/services/currency.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  };
+}
+
+export const currencyService: Service<CurrencyService> = {
+  name: "currency",
+  dependencies: ["localization", "model"],
+  deploy: async (env: OdooEnv): Promise<CurrencyService> => {
+    const result = await env.services
+      .model("res.currency")
+      .searchRead([], ["name", "symbol", "position", "decimal_places"]);
```
Comment (brboi) [inline addons/wowl/static/src/services/currency.ts]: get from session_info ?

File: addons/wowl/static/src/services/currency.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        id: c.id,
+        name: c.name,
+        symbol: c.symbol,
+        position: c.position,
+        digits,
+      };
+      currencies[currency.id] = currency;
+      currencies[currency.name] = currency;
+    }
+    return makeCurrencyService(env, currencies);
```
Comment (brboi) [inline addons/wowl/static/src/services/currency.ts]: remove makeCurrency... and no need to mock anymore

File: addons/wowl/static/src/services/localization.ts [hunk 4]
```diff
@@ -78,7 +168,115 @@ function parseDateTime(value: string): DateTime {
...
+  return function humanNumber(
+    number: number,
+    decimals: number = 0,
+    minDigits: number = 1,
+    formatterCallback: (s: string) => string = insertThousandsSep
+  ): string {
+    number = Math.round(number);
+
+    var d2 = Math.pow(10, decimals);
+    var val = _t("kMGTPE");
```
Comment (brboi) [inline addons/wowl/static/src/services/localization.ts]: ask mat for translations

File: addons/wowl/static/src/services/currency.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+export type Currency = {
+  id: number;
+  name: string;
+  symbol: string;
+  position: "after" | "before";
+  digits: [number, number];
+};
+
+export interface Currencies {
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/currency.ts]: remove this

File: addons/wowl/static/src/services/currency.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  [id: number]: Currency;
+  [name: string]: Currency;
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/currency.ts]: remove one of them here

File: addons/wowl/static/src/services/currency.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+      }
+      const ws = forceString ? " " : "&nbsp;";
+      if (currency.position === "after") {
+        return formatted_value + ws + currency.symbol;
+      } else {
+        return currency.symbol + ws + formatted_value;
+      }
+    };
+
+    const parse: CurrencyService["parse"] = (value, c) => {
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/currency.ts]: remove this

File: addons/wowl/static/src/services/currency.ts [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    const { currencies } = odoo.session_info;
+
+    const getAll: CurrencyService["getAll"] = () => Object.values(currencies);
+    const get: CurrencyService["get"] = (cid) => {
+      if (typeof cid === "number") {
+        return currencies[cid];
+      }
+      return getAll().find((c) => c.name === cid);
+    };
+
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/currency.ts:57]: const formatFloat = env.services.localization.formatFloat;

File: addons/wowl/static/tests/services/currency_tests.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+QUnit.module("Currency");
+
+QUnit.test("get & getAll", async (assert) => {
+  serviceRegistry = new Registry();
+  serviceRegistry.add("currency", currencyService);
+  env = await makeTestEnv({ serviceRegistry });
+  const { currency: curSvc } = env.services;
+
+  assert.deepEqual(curSvc.get(1)?.name, "USD");
```
Comment (ged-odoo) [inline addons/wowl/static/tests/services/currency_tests.ts]: `!.`

File: addons/wowl/static/tests/services/currency_tests.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
+import { currencyService } from "../../src/services/currency";
+import { makeTestEnv, OdooEnv } from "../helpers/index";
+import { Registries } from "../../src/types";
+import { Registry } from "../../src/core/registry";
```
Comment (ged-odoo) [inline addons/wowl/static/tests/services/currency_tests.ts:4]: readd old format monetary test if necessary

File: addons/wowl/static/tests/services/currency_tests.ts [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    curSvc.format(1234567.654, "EUR", { humanReadable: (value: number) => value > 100000 }),
+    "1M&nbsp;€"
+  );
+  assert.deepEqual(
+    curSvc.format(1234567.654, "EUR", { humanReadable: (value: number) => value < 100000 }),
+    "1234567.65&nbsp;€"
+  );
+});
+
+QUnit.test("parse", async (assert) => {
```
Comment (ged-odoo) [inline addons/wowl/static/tests/services/currency_tests.ts]: remove them

Comment (ged-odoo) [reply addons/wowl/static/tests/services/currency_tests.ts]: or comment them instead

File: addons/wowl/static/src/services/localization.ts [hunk 12]
```diff
@@ -1,88 +1,149 @@
...
- * Default values for localization parameters
- */
-export function getDefaultLocalization(): Localization {
-  return {
-    dateFormat: "%m/%d/%Y",
-    decimalPoint: ".",
-    direction: "ltr",
-    formatDateTime: makeFormatDateTime(strftimeToLuxonFormat("%m/%d/%Y %H:%M:%S")),
-    grouping: [],
-    langDateFormat: strftimeToLuxonFormat("%m/%d/%Y"),
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/localization.ts:35]: make 3 constants

File: addons/wowl/static/src/services/localization.ts [hunk 13]
```diff
@@ -166,7 +185,207 @@ _t("%d months ago");
...
+    }
+    number = sign * number;
+    return formatterCallback("" + number) + symbol;
+  };
+
+  const formatFloat: Localization["formatFloat"] = (value, field, options = {}) => {
+    if (value === false) {
+      return "";
+    }
+    if (
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/localization.ts]: remove this...

Comment (ged-odoo) [reply addons/wowl/static/src/services/localization.ts]: and remove them from options

File: addons/wowl/static/src/services/localization.ts [hunk 14]
```diff
@@ -166,7 +185,207 @@ _t("%d months ago");
...
+      // Luxon is not permissive regarding non alphabetical characters for
+      // formatting strings. So if the value to parse has less characters than
+      // the format, we would try to parse without the separators characters.
+      const woSeps = value.length < format.length && {
+        val: value.replace(/\D/g, ""),
+        fmt: format.replace(/\W/g, ""),
+      };
+
+      // FYI, luxon authorizes years until 275760 included...
+      const check = (d: DateTime): false | DateTime => d.isValid && d.year < 10000 && d;
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/localization.ts]: get this function out

File: addons/wowl/static/src/services/localization.ts [hunk 15]
```diff
@@ -166,7 +185,207 @@ _t("%d months ago");
...
+    const negative = num[0] === "-";
+    num = negative ? num.slice(1) : num;
+    return (negative ? "-" : "") + intersperse(num, langParams.grouping, langParams.thousandsSep);
+  };
+
+  const parseDate: Localization["parseDate"] = (value, options) => {
+    return parseDateTime(value, Object.assign({ dateOnly: true }, options));
+  };
+
+  const parseDateTime: Localization["parseDateTime"] = (value, options = {}) => {
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/localization.ts]: move this into utils date

File: addons/wowl/static/src/services/localization.ts [hunk 16]
```diff
@@ -166,7 +185,207 @@ _t("%d months ago");
...
+
+  Object.setPrototypeOf(translatedTerms, config.terms);
+
+  const langDateFormat = strftimeToLuxonFormat(langParams.dateFormat);
+  const langDateTimeFormat = strftimeToLuxonFormat(
+    `${langParams.dateFormat} ${langParams.timeFormat}`
+  );
+  const langTimeFormat = strftimeToLuxonFormat(langParams.timeFormat);
+
+  const insertThousandsSep: Localization["insertThousandsSep"] = (num) => {
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/localization.ts]: do not export this

File: addons/wowl/static/src/services/localization.ts [hunk 17]
```diff
@@ -166,7 +185,207 @@ _t("%d months ago");
...
+      result = result.minus({ minutes: result.offset });
+    }
+    result.toJSON = function () {
+      return this.setLocale("en").toFormat("yyyy-MM-dd HH:mm:ss");
+    };
+    return result;
+  };
+
+  const parseNumber: Localization["parseNumber"] = (value) => {
+    value = value.replace(new RegExp(escapeRegExp(langParams.thousandsSep), "g"), "");
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/localization.ts]: move regexp creation out of this

File: addons/wowl/static/src/services/localization.ts [hunk 18]
```diff
@@ -1,88 +1,149 @@
 import { DateTime } from "luxon";
-import { Service } from "../types";
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/localization.ts:2]: 1. move format and parse functions into utils/
2. use them in services

File: addons/wowl/static/src/utils/strings.ts [hunk 19]
```diff
@@ -13,3 +71,24 @@ export function sprintf(s: string, ...values: string[] | [{ [key: string]: strin
...
+ * Example:
+ *  - "aa-bb-CCcc-ddD xxxx-Yy-ZZ" -> "a-b-Cc-dD x-Yy-Z"
+ *  - "aa-bb-CCcc-ddD xxxx-Yy-ZZ" (ignoreCase: true) -> "a-b-C-d x-Y-Z"
+ *
+ * @param {String} str
+ * @param {Object} options
+ * @param {boolean} options.ignoreCase default: false
+ * @returns String
+ */
+export const stripAlphaDupes = (
```
Comment (ged-odoo) [inline addons/wowl/static/src/utils/strings.ts]: make this private inside the file that needs it

File: addons/wowl/static/tests/services/localization_tests.ts [hunk 20]
```diff
@@ -5,14 +5,20 @@ import { useService } from "../../src/core/hooks";
 import { getFixture, makeTestEnv, mount } from "../helpers/index";
 import { Registry } from "../../src/core/registry";
 import { makeFakeLocalizationService } from "../helpers/mocks";
+import { DateTime, Settings } from "luxon";
+import { LocalizationConfig } from "../../src/services/localization";
 
 let target: HTMLElement;
 let env: OdooEnv;
```
Comment (ged-odoo) [inline addons/wowl/static/tests/services/localization_tests.ts]: remove this and define env in each test

File: addons/wowl/static/src/services/localization.ts (hunk 12 above)
Comment (brboi) [reply addons/wowl/static/src/services/localization.ts:35]: in fact I already did :-)

https://github.com/odoo-dev/odoo/pull/538/files/998c7441fd92e6b12a2f1a6294edac69accab795..d1b853939b367f4b1bf5bdc368e855b38898c845#diff-4bfeeb6890f4e24db244e1da15796e0c036314900a0c194ff43e404d8eba6723R205-R209

File: addons/wowl/static/tests/services/localization_tests.ts [hunk 21]
```diff
@@ -5,51 +5,320 @@ import { useService } from "../../src/core/hooks";
...
+  assert.equal(date1.toISO(), date2.toISO(), "Day/month inverted + month i18n");
+
+  Settings.defaultLocale = originalLocale;
+});
+
+QUnit.skip("parseDate without separator", async (assert) => {
+  const env = await makeTestLocalizationEnv({ langParams: { dateFormat: "%d.%m/%Y" } });
+  const { localization: l10n } = env.services;
+  const dateFormat = "dd.MM/yyyy";
+  assert.throws(
```
Comment (ged-odoo) [inline addons/wowl/static/tests/services/localization_tests.ts]: unskip+ add comment/assertion

**Pull request merged**